docs/topics/compatibility-guide-18.md

```diff
@@ -0,0 +1,623 @@
+> - 1.5.30: fix the downcast behavior in generated property delegate accessors,
+>   `-Xuse-old-backend` can be used to temporarily revert to the pre-1.5.30 fix behavior
+> - \>= 1.9: fix the downcast behavior in other affected cases
+
+### Make private constructors of sealed classes really private
+
+> **Issue**: [KT-44866](https://youtrack.jetbrains.com/issue/KT-44866)
+>
+> **Component**: Core language
+>
+> **Incompatible change type**: source
+>
+> **Short summary**: After relaxing restrictions on where the inheritors of sealed classes can be declared in the project structure,
+> the default visibility of sealed class constructors became protected. However, until 1.8, Kotlin still allowed calling
+> explicitly declared private constructors a sealed class outside the class scope.
+>
+> **Deprecation cycle**:
+>
+> - 1.6.20: report a warning (or an error in the progressive mode) when private constructor of a sealed class 
+>   is called outside of that class
+> - 1.8.0: use default visibility rules for private constructors (call can be resolved to it only if it is inside corresponding class),
+>   the old behavior can be brought back temporarily by specifying `-XXLanguage:-UseConsistentRulesForPrivateConstructorsOfSealedClasses`
+>   compiler argument
+
+### Prohibit using operator `==` on incompatible numeric types in builder inference context
+
+> **Issue**: [KT-45508](https://youtrack.jetbrains.com/issue/KT-45508)
+>
+> **Component**: Core language
+>
+> **Incompatible change type**: source
+>
+> **Short summary**: Kotlin 1.8 will prohibit using the operator `==` on incompatible numeric types, for example, `Int` and `Long`,
+> in the scope of builder inference lambda functions, the same way as it does currently in other contexts
+>
+> **Deprecation cycle**:
+>
+> - 1.6.20: report a warning (or an error in the progressive mode) when the operator `==` is used on incompatible numeric types
+> - 1.8.0: raise the warning to an error,
+>   `-XXLanguage:-ProperEqualityChecksInBuilderInferenceCalls` can be used to temporarily revert to the pre-1.8 behavior
+
+### Prohibit `if` without `else` and non-exhaustive `when` in right hand side of elvis operator
+
+> **Issue**: [KT-44705](https://youtrack.jetbrains.com/issue/KT-44705)
+>
+> **Component**: Core language
+>
+> **Incompatible change type**: source
+>
+> **Short summary**: Kotlin 1.8 will prohibit using `if` operator without `else` branch or non-exhaustive `when`
+> in right hand side of elvis operator (`?:`). Previously, it was allowed if the elvis operator result was not used 
+> as an expression
+>
+> **Deprecation cycle**:
+>
+> - 1.6.20: report a warning (or an error in the progressive mode) on such non-exhaustive if and when expressions
+> - 1.8.0: raise this warning to an error,  
+>   `-XXLanguage:-ProhibitNonExhaustiveIfInRhsOfElvis` can be used to temporarily revert to the pre-1.8 behavior
+
+### Prohibit upper bound violation in a generic typealias usage
```

Comment:
```suggestion
### Prohibit upper bound violation in a generic type alias usage (several type arguments of the aliased type)
```

Reply:
I would show the difference in headings, otherwise, the first impression is that it's a mistake and we occasionally copy-pasted the same.

Reply:
```suggestion
### Prohibit upper bound violation in a generic type alias usage (one type parameter used in several type arguments of the aliased type)
```